File: androidx/src/main/java/com/androidx/media/MimeType.java
package com.androidx.media;

import android.os.Build;
import android.text.TextUtils;
import android.webkit.MimeTypeMap;

import com.androidx.utils.LegacyMimeUtils;

/**
 * user author: didikee
 * create time: 2019-12-03 11:38
 * description:
 *
 * mimetype列表参考：<a href="https://developer.android.com/reference/androidx/media3/common/MimeTypes">MimeTypes列表</a>
 */
public final class MimeType {
    public static final String UNKNOWN = "";
    public static final String ALL = "*/*";
    public static final String PREFIX_VIDEO = "video";
    public static final String PREFIX_AUDIO = "audio";
    public static final String PREFIX_IMAGE = "image";
    // Text
    public static final String TEXT = "text/plain";
    // image
    public static final String IMAGE = "image/*";
    public static final String PNG = "image/png";
    public static final String JPEG = "image/jpeg";
    public static final String GIF = "image/gif";
    public static final String WEBP = "image/webp";
    public static final String HEIC = "image/heic";
    public static final String HEIF = "image/heif";
    // video
    public static final String VIDEO = "video/*";
    public static final String MP4 = "video/mp4";
    public static final String _3GP = "video/3gp";
    // audio
    public static final String AUDIO = "audio/*";
    public static final String MP3 = "audio/mpeg";
    public static final String AAC = "audio/aac";
    public static final String _3GPP = "audio/3gpp";
    public static final String WAV = "audio/x-wav";
    public static final String AMR = "audio/amr";
    public static final String M4A = "audio/mpeg";

    // Document files
    public static final String PDF = "application/pdf";

    public static boolean isVideo(String mimeType) {
        return !TextUtils.isEmpty(mimeType) && mimeType.toLowerCase().startsWith(PREFIX_VIDEO);
    }

    public static boolean isImage(String mimeType) {
        return !TextUtils.isEmpty(mimeType) && mimeType.toLowerCase().startsWith(PREFIX_IMAGE);
    }

    public static boolean isAudio(String mimeType) {
        return !TextUtils.isEmpty(mimeType) && mimeType.toLowerCase().startsWith(PREFIX_AUDIO);
    }

    /**
     * 根据类型，获取对应的拓展名
     * 核心类，已经被隐藏无法直接查看
     * http://androidxref.com/4.4.4_r1/xref/libcore/luni/src/main/java/libcore/net/MimeUtils.java
     *
     * @param extension
     * @return
     */
    public static String getMimeTypeFromExtension(String extension) {
        String mimeTypeFromExtension;
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
            mimeTypeFromExtension = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
        } else {
            mimeTypeFromExtension = LegacyMimeUtils.guessMimeTypeFromExtension(extension);
        }
        if (TextUtils.isEmpty(mimeTypeFromExtension)) {
            return MimeType.UNKNOWN;
        } else {
            return mimeTypeFromExtension;
        }
    }

    /**
     * 根据拓展名，或者类型名
     * @param mimeType
     * @return
     */
    public static String getExtensionFromMimeType(String mimeType) {
        String extensionFromMimeType;
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
            extensionFromMimeType = MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType);
        } else {
            extensionFromMimeType = LegacyMimeUtils.guessExtensionFromMimeType(mimeType);
        }
        if (TextUtils.isEmpty(extensionFromMimeType)) {
            return "";
        } else {
            return extensionFromMimeType;
        }
    }

    public static String getMimeTypeFromFilename(String filename) {
        String extension = "";
        try {
            extension = filename.substring(filename.lastIndexOf(".") + 1);
        } catch (Exception e) {
            e.printStackTrace();
        }
        if (TextUtils.isEmpty(extension)) {
            return UNKNOWN;
        } else {
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT) {
                // aac要单独判断，因为4.4没有aac格式
                if ("aac".equalsIgnoreCase(extension)) {
                    return AAC;
                }
            }
            return getMimeTypeFromExtension(extension.toLowerCase());
        }
    }

}

File: androidx/src/main/java/com/androidx/media/MimeType.kt
package com.androidx.media

import android.os.Build
import android.text.TextUtils
import android.webkit.MimeTypeMap
import com.androidx.utils.LegacyMimeUtils
import java.util.Locale

/**
 * user author: didikee
 * create time: 2019-12-03 11:38
 * description:
 *
 * mimetype列表参考：[MimeTypes列表](https://developer.android.com/reference/androidx/media3/common/MimeTypes)
 */
object MimeType {
    const val UNKNOWN: String = ""
    const val ALL: String = "*/*"
    const val PREFIX_VIDEO: String = "video"
    const val PREFIX_AUDIO: String = "audio"
    const val PREFIX_IMAGE: String = "image"

    // Text
    const val TEXT: String = "text/plain"

    // image
    const val IMAGE: String = "image/*"
    const val PNG: String = "image/png"
    const val JPEG: String = "image/jpeg"
    const val GIF: String = "image/gif"
    const val WEBP: String = "image/webp"
    const val HEIC: String = "image/heic"
    const val HEIF: String = "image/heif"
    const val AVIF: String = "image/avif"

    // video
    const val VIDEO: String = "video/*"
    const val MP4: String = "video/mp4"
    const val _3GP: String = "video/3gp"

    // audio
    const val AUDIO: String = "audio/*"
    const val MP3: String = "audio/mpeg"
    const val AAC: String = "audio/aac"
    const val _3GPP: String = "audio/3gpp"
    const val WAV: String = "audio/x-wav"
    const val AMR: String = "audio/amr"
    const val M4A: String = "audio/mpeg"
    const val AC3:String = "audio/ac3"
    const val WMA:String = "audio/x-ms-wma"
    const val FLAC:String = "audio/flac"
    const val OGG:String = "audio/ogg"

    // Document files
    const val PDF: String = "application/pdf"

    @JvmStatic
    fun isVideo(mimeType: String?): Boolean {
        return !TextUtils.isEmpty(mimeType) && mimeType!!.lowercase(Locale.getDefault()).startsWith(
            PREFIX_VIDEO
        )
    }

    @JvmStatic
    fun isImage(mimeType: String?): Boolean {
        return !TextUtils.isEmpty(mimeType) && mimeType!!.lowercase(Locale.getDefault()).startsWith(
            PREFIX_IMAGE
        )
    }

    @JvmStatic
    fun isAudio(mimeType: String?): Boolean {
        return !TextUtils.isEmpty(mimeType) && mimeType!!.lowercase(Locale.getDefault()).startsWith(
            PREFIX_AUDIO
        )
    }

    /**
     * 根据类型，获取对应的拓展名
     * 核心类，已经被隐藏无法直接查看
     * http://androidxref.com/4.4.4_r1/xref/libcore/luni/src/main/java/libcore/net/MimeUtils.java
     *
     * @param extension
     * @return
     */
    @JvmStatic
    fun getMimeTypeFromExtension(extension: String?): String {
        val mimeTypeFromExtension = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
            MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
        } else {
            LegacyMimeUtils.guessMimeTypeFromExtension(extension)
        }
        return if (TextUtils.isEmpty(mimeTypeFromExtension)) {
            UNKNOWN
        } else {
            mimeTypeFromExtension!!
        }
    }

    /**
     * 根据拓展名，或者类型名
     * @param mimeType
     * @return
     */
    fun getExtensionFromMimeType(mimeType: String?): String {
        val extensionFromMimeType = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
            MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType)
        } else {
            LegacyMimeUtils.guessExtensionFromMimeType(mimeType)
        }
        return if (TextUtils.isEmpty(extensionFromMimeType)) {
            ""
        } else {
            extensionFromMimeType!!
        }
    }

    @JvmStatic
    fun getMimeTypeFromFilename(filename: String): String {
        var extension = ""
        try {
            extension = filename.substring(filename.lastIndexOf(".") + 1)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        if (TextUtils.isEmpty(extension)) {
            return UNKNOWN
        } else {
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT) {
                // aac要单独判断，因为4.4没有aac格式
                if ("aac".equals(extension, ignoreCase = true)) {
                    return AAC
                }
            }
            return getMimeTypeFromExtension(extension.lowercase(Locale.getDefault()))
        }
    }
}
